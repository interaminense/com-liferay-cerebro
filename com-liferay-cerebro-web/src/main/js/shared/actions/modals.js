export const actionTypes = {
	CLOSE_ALL_MODALS: 'CLOSE_ALL_MODALS',
	CLOSE_MODAL: 'CLOSE_MODAL',
	OPEN_MODAL: 'OPEN_MODAL'
};

export const modalTypes = {
	CREATE_MAPPING_MODAL: 'CREATE_MAPPING_MODAL',
	CSV_PREVIEW_MODAL: 'CSV_PREVIEW_MODAL',
	EDIT_CARD_MODAL: 'EDIT_CARD_MODAL',
	LOADING_MODAL: 'LOADING_MODAL',
	MATCH_FIELDS_MODAL: 'MATCH_FIELDS_MODAL',
	SELECT_ITEMS_MODAL: 'SELECT_ITEMS_MODAL',
	TEST: 'TEST'
};

export function close() {
	return {
		type: actionTypes.CLOSE_MODAL
	};
}

export function closeAll() {
	return {
		type: actionTypes.CLOSE_ALL_MODALS
	};
}

export function open(type, props = {}, options = {}) {
	const {closeOnBlur = true} = options;

	return {
		payload: {
			closeOnBlur,
			props,
			type
		},
		type: actionTypes.OPEN_MODAL
	};
}