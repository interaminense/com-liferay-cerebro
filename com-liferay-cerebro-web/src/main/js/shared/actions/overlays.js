export const actionTypes = {
	HIDE_OVERLAY: 'HIDE_OVERLAY',
	SHOW_OVERLAY: 'SHOW_OVERLAY'
};

export function hide(id) {
	return {
		payload: {
			id
		},
		type: actionTypes.HIDE_OVERLAY
	};
}

export function show(options) {
	const {
		alignment,
		containerClass,
		context,
		forceAlignment,
		id,
		offset,
		onMouseEnter,
		onMouseLeave,
		onOutsideClick,
		parentId
	} = options;

	return {
		payload: {
			alignment,
			containerClass,
			context,
			forceAlignment,
			id,
			offset,
			onMouseEnter,
			onMouseLeave,
			onOutsideClick,
			parentId
		},
		type: actionTypes.SHOW_OVERLAY
	};
}