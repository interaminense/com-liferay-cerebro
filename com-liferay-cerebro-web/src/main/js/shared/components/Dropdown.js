import autobind from 'autobind-decorator';
import Component, {Config} from 'metal-jsx';
import getCN from 'classnames';

import Button from './Button';
import Icon from './Icon';
import Overlay, {ALIGNMENTS} from './Overlay';

const ALIGNMENTS_ICON_MAP = {
	bottomCenter: 'caret-bottom',
	bottomLeft: 'caret-bottom',
	bottomRight: 'caret-bottom',
	leftCenter: 'caret-top',
	rightCenter: 'caret-top',
	topCenter: 'caret-top',
	topLeft: 'caret-top',
	topRight: 'caret-top'
};

class Section extends Component {
	render() {
		return <div class="dropdown-section">{this.props.children}</div>;
	}
}

class Subheader extends Component {
	render() {
		return <div class="dropdown-subheader">{this.props.children}</div>;
	}
}

class Item extends Component {
	@autobind
	handleClick() {
		const {hideOnClick, onClick} = this.props;

		onClick();

		if (hideOnClick && this.context.dropdown) {
			this.context.dropdown.hideMenu();
		}
	}

	render() {
		const {active, children, href} = this.props;

		const classes = getCN('dropdown-item', 'text-truncate', {active});

		return do {
			if (href) {
				<a {...this.otherProps()} class={classes} href={href}>
					{this.props.children}
				</a>;
			} else {
				<Button
					{...this.otherProps()}
					elementClasses={classes}
					onClick={this.handleClick}
				>
					{children}
				</Button>;
			}
		};
	}
}

Item.PROPS = {
	active: Config.bool().value(false),
	hideOnClick: Config.bool().value(false),
	href: Config.string(),
	onClick: Config.func()
};

class Dropdown extends Component {
	getChildContext() {
		return {
			dropdown: {
				hideMenu: this.handleToggle
			}
		};
	}

	@autobind
	handleOutsideClick() {
		this.state.active = false;
	}

	@autobind
	handleToggle() {
		this.state.active = !this.state.active;
	}

	render() {
		const {
			props: {
				align,
				buttonProps,
				caretDouble,
				children,
				disabled,
				elementClasses,
				forceAlignment,
				icon,
				label,
				readOnly,
				showCaret,
				toggleClasses
			},
			state: {active}
		} = this;

		const buttonClasses = getCN(
			'dropdown-toggle',
			{'nav-link': this.context.navbar},
			toggleClasses
		);

		const classes = getCN('dropdown', 'dropdown-root', elementClasses);

		return (
			<Overlay
				active={active}
				alignment={align}
				containerClass={classes}
				forceAlignment={forceAlignment}
				onOutsideClick={this.handleOutsideClick}
			>
				<div {...this.otherProps()} class={classes}>
					<Button
						{...buttonProps}
						disabled={disabled}
						elementClasses={buttonClasses}
						onClick={this.handleToggle}
					>
						<span class="text-truncate">
							{label && `${label} `}
						</span>

						{icon && (
							<span>
								<Icon symbol={icon} />
							</span>
						)}

						{!readOnly &&
							!icon &&
							showCaret && (
								<span>
									<Icon
										symbol={
											caretDouble
												? 'caret-double'
												: ALIGNMENTS_ICON_MAP[align]
										}
									/>
								</span>
							)}
					</Button>
				</div>

				<div class="dropdown-menu show">{children}</div>
			</Overlay>
		);
	}
}

Dropdown.PROPS = {
	align: Config.oneOf(ALIGNMENTS).value('bottomLeft'),
	buttonProps: Config.object().value({}),
	caretDouble: Config.bool(),
	disabled: Config.bool().value(false),
	forceAlignment: Config.bool().value(false),
	icon: Config.string().value(''),
	label: Config.oneOfType([Config.string(), Config.number()]),
	readOnly: Config.bool(),
	showCaret: Config.bool().value(true),
	toggleClasses: Config.string()
};

Dropdown.STATE = {
	active: Config.bool().value(false)
};

Dropdown.ALIGNMENTS = ALIGNMENTS;

Dropdown.Item = Item;
Dropdown.Section = Section;
Dropdown.Subheader = Subheader;

export default Dropdown;