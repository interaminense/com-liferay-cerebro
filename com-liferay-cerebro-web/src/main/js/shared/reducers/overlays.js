import {createReducer} from 'redux-toolbox';
import {OrderedMap, Map} from 'immutable';

import {actionTypes} from '../actions/overlays';

export default createReducer(OrderedMap(), {
	[actionTypes.HIDE_OVERLAY](state, action) {
		return state.delete(action.payload.id);
	},

	[actionTypes.SHOW_OVERLAY](state, action) {
		const {
			alignment,
			containerClass,
			context,
			forceAlignment,
			id,
			offset,
			onMouseEnter,
			onMouseLeave,
			onOutsideClick,
			parentId
		} = action.payload;

		return state.set(
			id,
			Map({
				alignment,
				containerClass,
				context,
				forceAlignment,
				id,
				offset,
				onMouseEnter,
				onMouseLeave,
				onOutsideClick,
				parentId
			})
		);
	}
});