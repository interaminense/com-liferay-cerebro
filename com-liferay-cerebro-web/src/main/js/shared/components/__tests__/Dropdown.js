jest.mock('shared/actions/overlays');

import * as actions from 'shared/actions/overlays';
import Dropdown from '../Dropdown';
import {renderWithStore} from 'test/mock-store';

describe('Dropdown', () => {
	let component;

	actions.show.mockReturnValue({type: 'NOOP'});
	actions.hide.mockReturnValue({type: 'NOOP'});

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render', () => {
		component = renderWithStore(Dropdown);

		expect(component).toMatchSnapshot();
	});

	it('should render without caret', () => {
		component = renderWithStore(Dropdown, {showCaret: false});

		expect(component).toMatchSnapshot();
	});

	it('should render with a different button display', () => {
		component = renderWithStore(Dropdown, {buttonProps: {display: 'link'}});

		expect(component).toMatchSnapshot();
	});

	it('should render items as links when href is passed to an item', () => {
		component = new Dropdown.Item({href: '#foo'});

		expect(component).toMatchSnapshot();
	});

	it('should open dropdown when handleToggle is called', () => {
		component = renderWithStore(Dropdown);

		expect(actions.show).not.toHaveBeenCalled();

		component.handleToggle();

		jest.runAllTimers();

		expect(actions.show).toHaveBeenCalled();
	});
});