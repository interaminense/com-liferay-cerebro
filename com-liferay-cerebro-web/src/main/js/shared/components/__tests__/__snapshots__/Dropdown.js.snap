// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 1`] = `
<div class="dropdown dropdown-root" id="overlay1">
  <button class="button-root btn btn-secondary dropdown-toggle" type="button">
    <span class="text-truncate"></span>
    <span>
      <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
      </svg>
    </span>
  </button>
</div>
`;

exports[`Dropdown should render items as links when href is passed to an item 1`] = `<a class="dropdown-item text-truncate" href="#foo"></a>`;

exports[`Dropdown should render with a different button display 1`] = `
<div class="dropdown dropdown-root" id="overlay3">
  <button class="button-root btn btn-link dropdown-toggle" type="button">
    <span class="text-truncate"></span>
    <span>
      <svg class="icon-root lexicon-icon lexicon-icon-caret-bottom">
        <use xlink:href="/images/lexicon/icons.svg#caret-bottom"></use>
      </svg>
    </span>
  </button>
</div>
`;

exports[`Dropdown should render without caret 1`] = `
<div class="dropdown dropdown-root" id="overlay2">
  <button class="button-root btn btn-secondary dropdown-toggle" type="button">
    <span class="text-truncate"></span>
  </button>
</div>
`;
